data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/stream/LogicalReplicationEventProcessor.java

```diff
@@ -0,0 +1,312 @@
+        } else if (messageType == MessageType.UPDATE.getValue()) {
+            processUpdateMessage(msg);
+        } else if (messageType == MessageType.DELETE.getValue()) {
+            processDeleteMessage(msg);
+        } else if (messageType == MessageType.COMMIT.getValue()) {
+            processCommitMessage(msg);
+        } else {
+            throw new IllegalArgumentException("Replication message type [" + messageType + "] is not supported. ");
+        }
+    }
+
+    void processBeginMessage(ByteBuffer msg) {
+        currentLsn = msg.getLong();
+        long epochMicro = msg.getLong();
+        currentEventTimestamp = convertPostgresEventTimestamp(epochMicro);
+        int transaction_xid = msg.getInt();
+
+        LOG.debug("Processed BEGIN message with LSN: {}, Timestamp: {}, TransactionId: {}", currentLsn, currentEventTimestamp, transaction_xid);
+    }
+
+    void processRelationMessage(ByteBuffer msg) {
+        int tableId = msg.getInt();
+        // null terminated string
+        String schemaName = getNullTerminatedString(msg);
+        String tableName = getNullTerminatedString(msg);
+        int replicaId = msg.get();
+        short numberOfColumns = msg.getShort();
+
+        List<String> columnNames = new ArrayList<>();
+        for (int i = 0; i < numberOfColumns; i++) {
+            int flag = msg.get();    // 1 indicates this column is part of the replica identity
+            // null terminated string
+            String columnName = getNullTerminatedString(msg);
+            ColumnType columnType = ColumnType.getByTypeId(msg.getInt());
+            String columnTypeName = columnType.getTypeName();
+            int typeModifier = msg.getInt();
+            if (columnType == ColumnType.VARCHAR) {
+                int varcharLength = typeModifier - 4;
+            } else if (columnType == ColumnType.NUMERIC) {
+                int precision = (typeModifier - 4) >> 16;
+                int scale = (typeModifier - 4) & 0xFFFF;
+            }
+            columnNames.add(columnName);
+        }
+
+        final List<String> primaryKeys = getPrimaryKeys(schemaName, tableName);
+        final TableMetadata tableMetadata = new TableMetadata(
+                tableName, schemaName, columnNames, primaryKeys);
+
+        tableMetadataMap.put((long) tableId, tableMetadata);
+
+        LOG.debug("Processed an Relation message with RelationId: {} Namespace: {} RelationName: {} ReplicaId: {}", tableId, schemaName, tableName, replicaId);
+    }
+
+    void processCommitMessage(ByteBuffer msg) {
+        int flag = msg.get();
+        long commitLsn = msg.getLong();
+        long endLsn = msg.getLong();
+        long epochMicro = msg.getLong();
+
+        if (currentLsn != commitLsn) {
+            // This shouldn't happen
+            pipelineEvents.clear();
+            throw new RuntimeException("Commit LSN does not match current LSN, skipping");
+        }
+
+        writeToBuffer(bufferAccumulator);
```

Comment:
QUES: will we flush uncommitted messages if numberOfRecordsToAccumulate or bufferTimeout is reached? If yes, Is it sth we want to avoid?

Reply:
This writeToBuffer() is called when processing the COMMIT message. So we will only write records to buffer when the changes are committed.

Reply:
Thanks for clarifying. One follow up question, could there be uncommitted operations stay in the bufferAccumulator forever since user has never commit it?